### PR TITLE
add some design docs.

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve? 
This PR opened some of internal design docs to `/docs`.
### Check List 

Tests
 - No code

### CODE REVIEW

File: docs/2019-09-09-BR-key-rewrite-disscussion.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    }
+    Cow::Borrowed(key)
+}
+```
+
+### Key Rewrite 對現時導入流程的影響
+
+Key Rewrite 對現時導入流程的影響
+現時無論是 BR 還是 Lightning，使用的都是同一套導入流程如下：
+1. 把 **KV 對**寫入到 RocksDB 實例（“**Engine File**”）來排序
```
Comment (5kbpers) [inline docs/2019-09-09-BR-key-rewrite-disscussion.md:67]: Please unify the traditional to simplified chinese.

File: docs/2019-08-05-new-design-of-backup-restore.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
+Last updated: 2020-03-07
```
Comment (5kbpers) [inline docs/2019-08-05-new-design-of-backup-restore.md:1]: I guess it could be better to rename the directory to `docs-cn`.

Comment (YuJuncen) [reply docs/2019-08-05-new-design-of-backup-restore.md:1]: How about `docs/cn`?

File: docs/2019-09-09-BR-key-rewrite-disscussion.md (hunk 1 above)
Comment (YuJuncen) [reply docs/2019-09-09-BR-key-rewrite-disscussion.md:67]: That would be good, but may spend some time, let me do it.

Comment (YuJuncen) [reply docs/2019-09-09-BR-key-rewrite-disscussion.md]: The translate was made by translate tools, with basic grammar check and term translation(like, 平行 -> 并行) by me. Some of term may still strange, review it if possible.

File: docs/cn/2019-08-05-new-design-of-backup-restore.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+> 这篇文档是 BR 的早期设计文档，大部分目标和现在相同；但是仍旧有着一些变动。
+>
+> 接下来使用 markdown 的“引用”格式来补充现在视角看上去的变化。
+
+背景：由于需要和 TiDB CDC 在技术上整合，旧版基于 learner 的备份实际上是不适用的，这个文档中描述的新方案基本思路是 kv scan。
+
+### 备份原理
+
+备份的原理其实就是 select *，但是这个方案和 mydumper 的区别：
```
Comment (kennytm) [inline docs/cn/2019-08-05-new-design-of-backup-restore.md:13]:
```suggestion
备份的原理其实就是 `select *`，但是这个方案和 mydumper 的区别：
```

File: docs/cn/2019-08-05-new-design-of-backup-restore.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+性能是 KV Scan 方案最头疼的问题之一，因为全表扫描势必会对线上业务造成影响。增量的实现也是全表扫，同样有性能问题，这个和传统数据库的增量不一样。好在也有优化的办法，比如学一下 crdb 在 SST 上加上 max ts 的 table property。
+
+### 外部存储
+
+备份存储设计了一个接口，下层可以有不同的实现，比如本地磁盘，s3 等。由于外部存储不一定是真正的磁盘，所以 TiKV 在生成 SST 的时候，不会直接把数据写到本地磁盘上，而是先缓存在内存中，生成完毕后直接写到外部存储中，这样能避免 IO，提高整个备份的吞吐，当然这里需要内存控制。
+
+### 异常处理
+
+异常处理备份期间的异常和 select * 一样，可分为两种可恢复和不可恢复，所有的异常都可以直接复用 TiDB 现有的机制。
```
Comment (kennytm) [inline docs/cn/2019-08-05-new-design-of-backup-restore.md:83]:
```suggestion
异常处理备份期间的异常和 `select *` 一样，可分为两种可恢复和不可恢复，所有的异常都可以直接复用 TiDB 现有的机制。
```

File: docs/cn/2019-08-05-new-design-of-backup-restore.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+2. 拿到新表的 table id，使用 table id 重写数据的 key，上传重写后 key-value 到 tikv-importer 上
+3. tikv-importer 接收到数据之后，重新切分、打散数据
+4. tikv-importer 将文件传到对应 tikv-server 上，随后调用 ingest 导入 SST 文件
+
+这个方案的优点是实现简单，不需要修改任何的现有代码，缺点是速度较慢，当 importer 在恢复过程中挂掉并且无法再启动时要从头开始恢复。
+
+### 初步对接计划
+
+恢复对接大体两部分：
+1. github.com/tikv/impoter
```
Comment (kennytm) [inline docs/cn/2019-08-05-new-design-of-backup-restore.md:147]:
```suggestion
1. <https://github.com/tikv/importer>
```

File: docs/cn/2019-08-05-new-design-of-backup-restore.md [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+  现阶段 importer 只能离线在全新集群上恢复，做不到在线恢复，这是不符合用户的预期的。这里的困境是 ingest SST 失败会导致各个副本之间的数据不一致。
+
+
+更新：已通过 [placement rule](https://pingcap.com/docs-cn/stable/how-to/configure/placement-rules/#placement-rules-使用文档) 实现资源隔离
+2. 不支持原集群恢复
+
+  原集群恢复涉及到各种 ID 的问题，最简单办法就是 rewrite key，把 key 中老 ID 替换成新 ID。这是 crdb 的方案，我们同样可以借鉴。
+
+更新：已通过 [Key rewrite](./2019-09-09-BR-key-rewrite-disscussion.md) 在理论上支持原集群恢复
```
Comment (kennytm) [inline docs/cn/2019-08-05-new-design-of-backup-restore.md:178]: These updates need to be placed in block quotes.

File: docs/cn/2019-08-05-new-design-of-backup-restore.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+### 附录
```
Comment (overvenus) [inline docs/cn/2019-08-05-new-design-of-backup-restore.md:180]: This section needs to be removed (from line 180 to line 214)

Comment (YuJuncen) [reply docs/cn/2019-08-05-new-design-of-backup-restore.md:180]: Should all the appendix be removed? Or just remove the comparison to "old solution" which is just discussed internally?

Comment (YuJuncen) [reply docs/cn/2019-08-05-new-design-of-backup-restore.md:180]: I just leave the link to pb-files, and remove all other part of the appendix.

File: docs/cn/2019-08-05-new-design-of-backup-restore.md (hunk 6 above)
Comment (YuJuncen) [reply docs/cn/2019-08-05-new-design-of-backup-restore.md]: That is done now.

File: docs/cn/2019-08-05-new-design-of-backup-restore.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
+Last updated: 2020-03-07
+
+## 新版备份恢复设计方案
+
+> 这篇文档是 BR 的早期设计文档，大部分目标和现在相同；但是仍旧有着一些变动。
```
[the remark this answers is not in the document]
Comment (YuJuncen) [reply docs/cn/2019-08-05-new-design-of-backup-restore.md:5]: Yes... I forgot it...

File: docs/cn/2019-08-05-new-design-of-backup-restore.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+> 事实上，现在没有做对 HTTP 的支持，除此之外，这里的“外部存储”还可以是本地的文件系统。
+5. TIKV 把执行信息汇报给外部程序。
+
+外部程序执行备份的基本流程：
+
+> 这一部分流程便是现在 BR 客户端中，也是未来的 TiDB 中所执行的流程。
+>
+> 不过现在我们会优先获得 schema，然后根据表的 Key Range 下推执行备份。
+
+1. 下推备份到所有 tikv
```
[the remark this answers is not in the document]
Comment (YuJuncen) [reply docs/cn/2019-08-05-new-design-of-backup-restore.md]: I made a commit that replace all tikv into TiKV(expect for `tikv-***`).

File: docs/cn/2019-09-09-BR-key-rewrite-disscussion.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+### Key Rewrite 对现在导入流程的影响
+
+Key Rewrite 对现在导入流程的影响
+现在无论是 BR 还是 Lightning，使用的都是同一套导入流程，如下：
+1. 把 **KV 对**写入到 RocksDB 实例（“**Engine File**”）来排序
+2. (Prepare) 遍历此 Engine file，
+   1. (Pre-split) 每当 K+V 的累积长度达 512 MB 时（可设置），对这个 **Range** 执行 PrepareRangeJob，如下。
+   2. (GetRegion) 执行 *PD* 的 *get_region_info* 取得这个 *Range.start* 对应的 *Region* 信息。
+   3. (Split) 把这个 *Region* 从 *Range.end* 处使用 *split_region_opt* 分裂成两部分。如果出 EpochNotMatch 或 NotLeader 的错会由 b 步开始重试。
```
Comment (3pointer) [inline docs/cn/2019-09-09-BR-key-rewrite-disscussion.md:63]:
```suggestion
   3. (Split) 把这个 *Region* 从 *Range.end* 处使用 *split_region_opt* 分裂成两部分。如果出 EpochNotMatch 或 NotLeader 的错会由 II 步开始重试。
```

**Comment by 3pointer** [comment]
I have two suggestions:
1. `docs/cn/2020-04-12-introduction-to-br.md`  looks like a introduction article, not a design  doc maybe we should give another PR and put it into our [Official docs](https://github.com/pingcap/docs-cn/blame/master/TOC.md#L61), and we can do it later.
2. `2019-09-24-BR-and-lightning-reorganization.md` seems not clear, and not finished yet, how about remove it?

Rest LGTM

**Comment by YuJuncen** [comment]
> I have two suggestions:
> 
> 1. `docs/cn/2020-04-12-introduction-to-br.md`  looks like a introduction article, not a design  doc maybe we should give another PR and put it into our [Official docs](https://github.com/pingcap/docs-cn/blame/master/TOC.md#L61), and we can do it later.
> 2. `2019-09-24-BR-and-lightning-reorganization.md` seems not clear, and not finished yet, how about remove it?
> 
> Rest LGTM

I think so. For introduction usage, the file `2019-08-05-new-design-of-backup-restore.md` is enough now. And it's more clear for developer looking for reference.

The architecture graph in `2019-09-24-BR-and-lightning-reorganization.md` maybe useful for some other document, so I leave it for possible usage like in README.

File: docs/cn/2019-09-09-BR-key-rewrite-disscussion.md (hunk 10 above)
Comment (YuJuncen) [reply docs/cn/2019-09-09-BR-key-rewrite-disscussion.md:63]: oops... thanks!

**Comment by YuJuncen** [comment]
/run-all-tests

```
Error: ingest error message:"region 27628 not found" region_not_found:<region_id:27628 > : ingest sst failed
```
Seems like #238 .

File: docs/cn/2019-09-09-BR-key-rewrite-disscussion.md [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    Cow::Borrowed(key)
+}
+```
+
+### Key Rewrite 对现在导入流程的影响
+
+现在无论是 BR 还是 Lightning，使用的都是同一套导入流程，如下：
+1. 把 **KV 对**写入到 RocksDB 实例（“**Engine File**”）来排序
+2. (Prepare) 遍历此 Engine file，
+   1. (Pre-split) 每当 K+V 的累积长度达 512 MB 时（可设置），对这个 **Range** 执行 PrepareRangeJob，如下。
```
Comment (overvenus) [inline docs/cn/2019-09-09-BR-key-rewrite-disscussion.md:60]: There are Ligthning only. Cc @3pointer

File: docs/cn/2019-09-09-BR-key-rewrite-disscussion.md [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+当执行了 Key Rewrite，上下游的 Key Range 就会不一致。我们使用**红色**来标示使用 Rewrite 前的 Keys 的对象、用*蓝色*标示 Rewrite 后的对象。其中不变的是：
+
+1. 源数据必然是 Rewrite 前的
+2. PD 和 Ingest 后的结果必然是 Rewrite 后的。
+
+我们看到 Rewrite 前后的 Keys 在各步骤交叉被使用。这里的问题是怎么选一个合适的位置去 Rewrite Keys 来为 BR 提取最大性能。
+
+### 解决方案：Key rewrite before ingest in every replica
+
+<img src="../resources/solution3-of-key-rewrite.svg" alt="img" style="zoom:50%;" />
```
Comment (overvenus) [inline docs/cn/2019-09-09-BR-key-rewrite-disscussion.md:81]:
```suggestion
<img src="../resources/solution3-of-key-rewrite.svg" alt="solution3-of-key-rewrite" width="500" />
```

File: docs/cn/2019-09-09-BR-key-rewrite-disscussion.md (hunk 11 above)
Comment (YuJuncen) [reply docs/cn/2019-09-09-BR-key-rewrite-disscussion.md:60]: This discussion is about tikv-importer, which would be replaced by the download & ingest API. That seems like, split region will be done at "client" side, both BR and lightning can have their own behavior.

As a result, should we replace description with current BR behavior?

File: docs/cn/2019-09-09-BR-key-rewrite-disscussion.md [hunk 13]
```diff
@@ -54,12 +54,12 @@ fn undo_rewrite_key(rules: &[RewriteRule], key: &[u8]) -> Cow<[u8]> {
 
 ### Key Rewrite 对现在导入流程的影响
 
-现在无论是 BR 还是 Lightning，使用的都是同一套导入流程，如下：
+现在 BR 的导入流程如下：
 1. 把 **KV 对**写入到 RocksDB 实例（“**Engine File**”）来排序
```
Comment (overvenus) [inline docs/cn/2019-09-09-BR-key-rewrite-disscussion.md:58]:
```suggestion
1. 把 **KV 对**写入到 RocksDB 实例（“**SST File**”）来排序
```

File: docs/cn/2019-09-09-BR-key-rewrite-disscussion.md [hunk 14]
```diff
@@ -54,12 +54,12 @@ fn undo_rewrite_key(rules: &[RewriteRule], key: &[u8]) -> Cow<[u8]> {
...
-2. (Prepare) 遍历此 Engine file，
-   1. (Pre-split) 每当 K+V 的累积长度达 512 MB 时（可设置），对这个 **Range** 执行 PrepareRangeJob，如下。
-   2. (GetRegion) 执行 *PD* 的 *get_region_info* 取得这个 *Range.start* 对应的 *Region* 信息。
-   3. (Split) 把这个 *Region* 从 *Range.end* 处使用 *split_region_opt* 分裂成两部分。如果出 EpochNotMatch 或 NotLeader 的错会由 GetRegion 开始重试。
+2. (Prepare) 在 client 侧遍历此 Engine file，
```
Comment (overvenus) [inline docs/cn/2019-09-09-BR-key-rewrite-disscussion.md:59]:
```suggestion
2. (Prepare) 在 client 侧遍历此 SST file，
```

**Pull request merged**